### PR TITLE
Parse SQLFiles with ANTLR grammar

### PR DESCRIPTION
**Pull request by frapontillo**
#### Problem

`SQLFile` parses the input statements manually, without relying on a proper state machine nor grammar definition. Therefore, it may discard syntactically valid statements such as:

```sql
CREATE TABLE `all--pizzas`(id INTEGER);
```

since it naively assumes that everything that comes after `--` is a line comment'.

Furthermore, it doesn't acknowledge multiple statements unless they are separated by a new line:

```sql
CREATE TABLE `one`(id INTEGER); CREATE TABLE `two`(id INTEGER);
```

#### Solution

This PR uses a SQLite grammar definition for ANTLR4 to address these issues (and potentially many more).

The initial grammar definition is, [per SQLite reference](https://www.sqlite.org/lang.html):

```
sql_stmt_list : sql_stmt (';' sql_stmt )* ';'
```

`SQLFile#parse` now accesses the parsing tree, then moves to `sql_stmt_list`, looping on all the contained `sql_stmt`.

The grammar lexer will also be leveraged in an upcoming PR for the `TableCreateStatementParser`.

##### Test(s) added 

Changed tests:

- do not remove line comments
- do not remove block comments

New tests:

- do not remove double dashes everywhere, they may be part of the statement as they aren't always a comment
- make sure that multiple statements on a single line are correctly returned separately

###### Paired with 

Noam Chomsky

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -7,7 +7,7 @@ buildscript {
         }
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:1.1.0'
+        classpath 'com.android.tools.build:gradle:2.0.0'
```
Comment (eduardb) [inline build.gradle]: Isn't 2.1.0 the latest stable version?

Comment (frapontillo) [inline build.gradle]: Shall we try it? 💃

Comment (frapontillo) [inline build.gradle]: Done in cbf36c7

Comment (eduardb) [inline build.gradle]: Probably gradle plugin updates should be done in a separate PR ( @blundell  would agree with this :trollface: ), but if we update it, better update it to the latest stable version

File: analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java [hunk 2]
```diff
@@ -1,96 +1,85 @@
 package com.novoda.sqlite.impl;
 
-import java.io.*;
```
Comment (eduardb) [inline analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java]: 👍

File: build.gradle (hunk 1 above)
Comment (frapontillo) [inline build.gradle]: Yeah, the problem is it wasn't even building with the older plugin, therefore Paul and I needed to do this here.

File: analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java [hunk 3]
```diff
@@ -1,96 +1,85 @@
...
-			file.parse(reader);
-			return file.getStatements();
-		} finally {
-			if (reader != null) {
-				reader.close();
-			}
-		}
-	}
+    private static final String NEW_LINE = "\n";
+    private static final String SEMICOLON = ";";
```
Comment (devisnik) [inline analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java]: I'd prefer to inline these, they don't add any value here, imho.

Comment (frapontillo) [inline analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java]: Addressed in 8d5c9378b818ac1ab61af952ea61c49ebfbe812c

File: analyzer/build.gradle [hunk 4]
```diff
@@ -35,6 +37,14 @@ version = '0.3.2'
 sourceCompatibility = JavaVersion.VERSION_1_6
 targetCompatibility = JavaVersion.VERSION_1_6
 
+sourceSets {
+    main.java.srcDir 'src/main/build'
+}
```
Comment (devisnik) [inline analyzer/build.gradle]: Is this needed? I'd expect the `antlr` plugin to take care of that.

Comment (frapontillo) [inline analyzer/build.gradle]: It doesn't 😞

File: analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java [hunk 5]
```diff
@@ -1,96 +1,82 @@
...
+            statementList.addAll(statements.sql_stmt());
+        }
+
+        this.statements = new ArrayList<String>(statementList.size());
+        for (SQLiteParser.Sql_stmtContext statement : statementList) {
+            int startIndex = statement.getStart().getStartIndex();
+            int stopIndex = statement.getStop().getStopIndex();
+            String textStatement = sqlContent.substring(startIndex, stopIndex + 1) + ";";
+            this.statements.add(textStatement);
+        }
```
Comment (devisnik) [inline analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java]: Might this method benefit from some smaller sub-methods?

**Comment by devisnik** [comment]
Awesome stuff. 👍

File: analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java (hunk 5 above)
Comment (frapontillo) [inline analyzer/src/main/java/com/novoda/sqlite/impl/SQLFile.java]: Done in 3d35e17278a35bb04f34fe156ab6c6d3dabd4626

File: settings.gradle [hunk 6]
```diff
@@ -1,2 +1,3 @@
-include 'demo-assethelper', 'demo-sqliteprovider'
-
+include 'demo-assethelper'
+include 'demo-sqliteprovider'
+include 'analyzer'
```
Comment (devisnik) [inline settings.gradle]: We didn't include the actual plugin project here. It's a separate Gradle project, linked into buildSrc of the root project via some magic in the buildSrc's `build.gradle`.

Comment (frapontillo) [inline settings.gradle]: That was probably working on previous AS versions, but if we don't include `analyzer` explicitly, it doesn't get  picked up as a module by Android Studio.

File: analyzer/build.gradle (hunk 4 above)
Comment (devisnik) [inline analyzer/build.gradle]: I just removed lines 40 to 46. Seems to work for me locally. 
What's the issues you're experiencing?

File: settings.gradle (hunk 6 above)
Comment (devisnik) [inline settings.gradle]: What I mean in the comment above is: line 3 should be removed.

Comment (frapontillo) [inline settings.gradle]: If I do that, this happens:

image

File: analyzer/build.gradle (hunk 4 above)
Comment (frapontillo) [inline analyzer/build.gradle]: It says it can't find the classes among generated code.

File: settings.gradle (hunk 6 above)
Comment (devisnik) [inline settings.gradle]: Yes, that's because it is a separate module. You can import just the `analyzer`'s build.gradle as a project by itself. 
Not ideal, but helps to link the plugin's sources to the sample projects.

I usually have the two projects open (in two different windows).

File: analyzer/build.gradle (hunk 4 above)
Comment (devisnik) [inline analyzer/build.gradle]: Inside AS, or when running from command line?

Comment (frapontillo) [inline analyzer/build.gradle]: Command line.

File: settings.gradle (hunk 6 above)
Comment (frapontillo) [inline settings.gradle]: Are there any drawbacks in adding the `analyzer` module in `settings.gradle`?

Comment (blundell) [inline settings.gradle]: I was confused about this as well. If you get it on the path or not on the path. Can you update the readme and add another readme to this module to explain this confusion or to explain why it's on the path but not actually depended on

File: analyzer/build.gradle (hunk 4 above)
Comment (devisnik) [inline analyzer/build.gradle]: Hm, weird, I'll do a PR against this branch for you to check.

Comment (devisnik) [inline analyzer/build.gradle]: see https://github.com/novoda/sqlite-analyzer/pull/50

File: settings.gradle (hunk 6 above)
Comment (devisnik) [inline settings.gradle]: @frapontillo If we add it as a regular module, that breaks the build. 

The problem is that a Gradle plugin is a buildscript dependency. As such it is resolved before any regular modules. Usually you'd need to deploy a new version of the plugin to a local repository and then let your samples depend on that. We managed to directly link the sources via some wiring in the `buildSrc` folder. Not pretty, but much faster than having to deploy each time.

Comment (frapontillo) [inline settings.gradle]: It doesn't break the build for me though. Does it break on your machine? 😱

File: analyzer/build.gradle (hunk 4 above)
Comment (frapontillo) [inline analyzer/build.gradle]: Sorry, I misread your comment. Yes, removing `generateGrammarSource` doesn't break anything.
Good catch! 💯 

Done in 3f4ffa519ccf16efac8b18d10df165cf6a66e7ad

File: settings.gradle (hunk 6 above)
Comment (devisnik) [inline settings.gradle]: Breaks for me if I call `assemble` in the `analyzer` module for example.

Comment (frapontillo) [inline settings.gradle]: I'm no Gradle expert, but could the break be caused by the gradle 2.3 in the `analyzer` directory? I have tried to update it to 2.10 and the build succeeds if launched from the `analyzer` folder.

Also, if we keep `analyzer` in `settings.gradle` we can remove the `analyzer/gradle` folder and gradlew files, right?

Comment (devisnik) [inline settings.gradle]: Yes, the breaking is related to the Gradle version check, which is caused by the `demo-assethelper` configuration. The thing is, building the plugin should not cause any demo projects to be rebuilt. 
We need to be able to test against different Gradle versions, independently of the version used to build the plugin.

See also https://github.com/novoda/sqlite-analyzer/pull/19

Comment (frapontillo) [inline settings.gradle]: Oh ok, now I understand the reason behind this 👍 . Can you please update your PR #50 with some readme changes to make this clear? I'll be happy to merge it in here before merging this.

Comment (devisnik) [inline settings.gradle]: Okay, let me have a look.

Comment (devisnik) [inline settings.gradle]: @frapontillo I've added a README to the plugin module in #50

Comment (frapontillo) [inline settings.gradle]: Done! 53ff074
